### PR TITLE
chore: adding initial localhost client testing

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Adds unit testing for 09-localhost client
- Fix verify membership/non-membership key paths to omit prefix

closes: #XXXX

### Commit Message / Changelog Entry

```bash
chore: adding initial localhost client testing
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/light-clients/09-localhost/client_state.go [hunk 1]
```diff
@@ -95,7 +91,17 @@ func (cs ClientState) VerifyMembership(
...
+	if !ok {
+		return sdkerrors.Wrapf(clienttypes.ErrFailedChannelStateVerification, "todo: update error -- not found for path %s", path)
+	}
+
+	// todo: revisit this
+	if len(merklePath.GetKeyPath()) != 2 && merklePath.KeyPath[0] != exported.StoreKey {
+		return sdkerrors.Wrapf(clienttypes.ErrFailedChannelStateVerification, "todo: update error -- not found for path %s", path)
+	}
+
+	bz := store.Get([]byte(merklePath.KeyPath[1]))
```
Comment (chatton) [inline modules/light-clients/09-localhost/client_state.go:104]: could be a good candidate for an unexported function.
```go
func readPathBytes(path exported.Path,) ([]byte, error) {...}
```

File: modules/light-clients/09-localhost/client_state.go [hunk 2]
```diff
@@ -122,7 +128,17 @@ func (cs ClientState) VerifyNonMembership(
...
```
Comment (chatton) [inline modules/light-clients/09-localhost/client_state.go:141]: could use the function I suggested above here too

File: modules/light-clients/09-localhost/client_state.go [hunk 3]
```diff
@@ -153,8 +169,10 @@ func (cs ClientState) UpdateStateOnMisbehaviour(_ sdk.Context, _ codec.BinaryCod
 func (cs ClientState) UpdateState(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg exported.ClientMessage) []exported.Height {
 	height := clienttypes.GetSelfHeight(ctx)
 
-	clientState := NewClientState(ctx.ChainID(), height)
-	clientStore.Set([]byte(host.KeyClientState), clienttypes.MustMarshalClientState(cdc, clientState))
+	cs.ChainId = ctx.ChainID()
+	cs.LatestHeight = height
+
+	clientStore.Set([]byte(host.KeyClientState), clienttypes.MustMarshalClientState(cdc, &cs))
```
Comment (chatton) [inline modules/light-clients/09-localhost/client_state.go:175]: just to make sure I understand the intention correctly, here we modify `cs` the method receiver, but it is not a pointer, so the CientState in the store will be updated, but the instance we act on i.e. 

```go
clientState.UpdateState(...)
```

will remain unchanged.

**Review by chatton — COMMENTED** [review]
Looking good so far, I'll take a more in depth look at the test cases once the PR is marked ready for review!

File: modules/light-clients/09-localhost/client_state.go (hunk 3 above)
Comment (damiannolan) [reply modules/light-clients/09-localhost/client_state.go:173]: yeah, correct, this is essentially the same as how the solomachine client state is doing updates also.

File: modules/light-clients/09-localhost/client_state.go (hunk 1 above)
Comment (damiannolan) [reply modules/light-clients/09-localhost/client_state.go]: I think its more or less ok as is, given its a relatively small method compared to solomachine or tm

**Comment by damiannolan** [comment]
Created https://github.com/cosmos/ibc-go/issues/3099 to address error returns in a follow-up PR.

File: modules/light-clients/09-localhost/client_state.go [hunk 4]
```diff
@@ -95,7 +91,16 @@ func (cs ClientState) VerifyMembership(
...
+	merklePath, ok := path.(commitmenttypes.MerklePath)
+	if !ok {
+		return sdkerrors.Wrapf(clienttypes.ErrFailedChannelStateVerification, "todo: update error -- not found for path %s", path)
+	}
+
+	if len(merklePath.GetKeyPath()) != 2 {
+		return sdkerrors.Wrapf(clienttypes.ErrFailedChannelStateVerification, "todo: update error -- not found for path %s", path)
+	}
+
+	bz := store.Get([]byte(merklePath.KeyPath[1]))
```
Comment (damiannolan) [inline modules/light-clients/09-localhost/client_state.go:103]: `merklePath.KeyPath` is a `[]string` where the first element of the key path `"/ibc/"` (the prefix that is applied [here](https://github.com/cosmos/ibc-go/blob/main/modules/core/03-connection/keeper/verify.go#L177)), but here we pass the core IBC store for verification, so we can trim the prefix or use `keyPath[1]` (i.e. the key path without the prefix)

Comment (chatton) [reply modules/light-clients/09-localhost/client_state.go:103]: I feel like we should include this reasoning in a comment alongside these checks.

**Review by chatton — APPROVED** [review]
Really nice work with these tests, and nice job catching the ibc prefix issue! 👍

File: modules/light-clients/09-localhost/client_state.go (hunk 4 above)
Comment (damiannolan) [reply modules/light-clients/09-localhost/client_state.go:104]: good call, done!

**Review by chatton — APPROVED** [review]
ack changes 👍

File: modules/light-clients/09-localhost/client_state.go (hunk 4 above)
Comment (AdityaSripal) [reply modules/light-clients/09-localhost/client_state.go:104]: Should we verify that the store is prefixed by KeyPath[0]? I don't know if there's an easy way to do that

File: modules/light-clients/09-localhost/client_state.go [hunk 5]
```diff
@@ -153,8 +169,10 @@ func (cs ClientState) UpdateStateOnMisbehaviour(_ sdk.Context, _ codec.BinaryCod
 func (cs ClientState) UpdateState(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg exported.ClientMessage) []exported.Height {
 	height := clienttypes.GetSelfHeight(ctx)
 
-	clientState := NewClientState(ctx.ChainID(), height)
-	clientStore.Set([]byte(host.KeyClientState), clienttypes.MustMarshalClientState(cdc, clientState))
+	cs.ChainId = ctx.ChainID()
```
Comment (AdityaSripal) [inline modules/light-clients/09-localhost/client_state.go:172]: Maybe should be a separate PR. But do we need the chainID if the function got removed above

**Review by AdityaSripal — APPROVED** [review]
